src/ol/expr/gpu.js

```diff
@@ -166,6 +166,13 @@ export const PALETTE_TEXTURE_ARRAY = 'u_paletteTextures';
 export const FEATURE_ID_PROPERTY_NAME = 'featureId';
 export const GEOMETRY_TYPE_PROPERTY_NAME = 'geometryType';
 
+/**
+ * The value `-9999999` will be used to indicate that a property on a feature is not defined, similar to a "no data" value.
+ * @type {number}
+ * @api
+ */
+export const UNDEFINED_PROP_VALUE = -9999999;
```

Comment:
Not 100% sure about this one but the alternative is to add another value in the vertex buffers and that has a big impact on memory consumption; I figure that using a fixed value like this would cause very limited side effects. If a property ends up being defined and having this value, the only side effect is that the `'has'` operator would return true for this property; expressions would still run normally.

Reply:
The export here does not allow users to change the value, so it does not make much sense to make it `@api`. In the docs, just mention the `-9999999` value instead.

Regarding the approach itself, it would be safer to compare against two values. I'd suggest `Infinity` and `-Infinity`. Both can be expressed as positive or negative `INF` and checked for with `isinf()` in glsl.

Reply:
Yes I thought about that, but unfortunately I don't think there's a way to write "infinity" in an array buffer. The "infinite" value in GLSL can appear in case of division by zero, but cannot be put in an attribute.

I made the const part of the API so that an application code can take it into account and work around it without writing down the magic number as is. But that might be giving too much importance to this const.

> Regarding the approach itself, it would be safer to compare against two values. I'd suggest Infinity and -Infinity, or a very high positive and negative number.

I didn't quite get that; how would I compare against two fixed values?

Reply:
If the above looks like overkill, you can stick with your approach, but I'd suggest at least adding a `console.warn` when the `UNDEFINED_PROP_VALUE` is used.

Reply:
I like the `console.warn` idea :) I'll add it.

Should I leave the magic number as part of the API? I'm tempted to remove it and see if the need for it comes up later on.

Reply:
I'd say remove it. Thanks!

Reply:
Added in https://github.com/openlayers/openlayers/pull/16717/commits/8ef19135b818b79e4698de4ecfadaefaa1a59b91